File: doc/codes/python/examples/tutorial/multiple-device.py
from scipy.spatial.transform import Rotation

from autd3.geometry import Autd3, Geometry

def main() -> None:
    # ANCHOR: translation
    Geometry(
        [
            Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0]),
            Autd3(
                origin=(Autd3.DEVICE_WIDTH, 0.0, 0.0),
                rotation=(1.0, 0.0, 0.0, 0.0),
            ),
        ]
    )
    # ANCHOR_END: translation

    # ANCHOR: global
    Geometry(
        [
            Autd3(
                origin=(Autd3.DEVICE_WIDTH, 0.0, 0.0),
                rotation=(1.0, 0.0, 0.0, 0.0),
            ),
            Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0]),
        ]
    )
    # ANCHOR_END: global

    # ANCHOR: rotation
    Geometry(
        [
            Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0]),
            Autd3(
                origin=(0.0, 0.0, Autd3.DEVICE_WIDTH),
                rotation=Rotation.from_euler("y", 90.0, degrees=True),
            ),
        ]
    )
    # ANCHOR_END: rotation


if __name__ == "__main__":
    main()

File: doc/codes/rust/examples/tutorial/multiple-device.rs
use std::f32::consts::FRAC_PI_2;

use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry, Point3, UnitQuaternion, Vector3};

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    // ANCHOR: translation
    Geometry::new(vec![
        Autd3::new(Point3::origin(), UnitQuaternion::identity()),
        Autd3::new(
            Point3::new(Autd3::DEVICE_WIDTH, 0.0, 0.0),
            UnitQuaternion::identity(),
        ),
    ]);
    // ANCHOR_END: translation

    // ANCHOR: global
    Geometry::new(vec![
        Autd3::new(
            Point3::new(-Autd3::DEVICE_WIDTH, 0.0, 0.0),
            UnitQuaternion::identity(),
        ),
        Autd3::new(Point3::origin(), UnitQuaternion::identity()),
    ]);
    // ANCHOR_END: global

    // ANCHOR: rotation
    Geometry::new(vec![
        Autd3::new(Point3::origin(), UnitQuaternion::identity()),
        Autd3::new(
            Point3::new(0.0, 0.0, Autd3::DEVICE_WIDTH),
            UnitQuaternion::from_axis_angle(&Vector3::y_axis(), FRAC_PI_2),
        ),
    ]);
    // ANCHOR_END: rotation

    Ok(())
}
